text/3371-cargo-target-base-dir.md

```diff
@@ -0,0 +1,402 @@
+
+*Note: Bazel 5.4.0 was used as the reference, the latest stable version as of this writing, things may change in the future or be different for older versions.*
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+<!--
+- What parts of the design do you expect to resolve through the RFC process before this gets merged?
+- What parts of the design do you expect to resolve through the implementation of this feature before stabilization?
+- What related issues do you consider out of scope for this RFC that could be addressed in the future independently of the solution that comes out of this RFC?
+-->
+
+- Do we want to add forward links for untemplated `CARGO_TARGET_DIR` too in the process of this RFC ?
+- Do we want to differentiate according to users ? `bazel` is a generic build tool, whereas `cargo` is not, so maybe differentiating on users is not necessary for the latter ?
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+<!--
+Think about what the natural extension and evolution of your proposal would
+be and how it would affect the language and project as a whole in a holistic
+way. Try to use this section as a tool to more fully consider all possible
+interactions with the project and language in your proposal.
+Also consider how this all fits into the roadmap for the project
+and of the relevant sub-team.
+
+This is also a good place to "dump ideas", if they are out of scope for the
+RFC you are writing but otherwise related.
+
+If you have tried and cannot think of any future possibilities,
+you may simply state that you cannot think of anything.
+
+Note that having something written down in the future-possibilities section
+is not a reason to accept the current or a future RFC; such notes should be
+in the section on motivation or rationale in this or subsequent RFCs.
+The section merely provides additional information.
+-->
+
+- Introduce remapping into the concept in some way.
+- Introduce a form of garbage collection. Expanded upon this [Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/last-use.20tracking).
+
+## Adding `XDG_CACHE_HOME`, `HOME` or `CARGO_HOME` as templates
+
+Not all OSes use the XDG convention, notably Windows and macOS (though the latter can be somewhat made to) and it is very easy to define `CARGO_TARGET_DIR=${XDG_CACHE_HOME:-~/.cache}/cargo-target-directories/{manifest-path-hash}` if wanted by users.
+
+It won't work in the `config.toml` but it will work with the environment variable and the command line option, both of which override the TOML config.
+
+It is certainly possible to add at least `{home}` and `{cargo_home}` but it can be done in the future without interfering at all with `{manifest-path-hash}`, making it relatively easy to defer and add if there is demand.
```

Comment:
We could also have a variable for "user target base directory" (ie where cargo script writes to)

Reply:
As a template for the target directory ? I don't understand.

If you're talking about where `cargo script` will have its own target directory to write binaries and such, `{manifest-path-hash}` is already enough: it targets the manifest and not the directory to handle scripts that are in the same directory, so they will have their own target directory

Reply:
We will have a canonical directory within `CARGO_HOME` (and wherever we move it when we do  rust-lang/cargo#1734) for target directories.  It would make sense to provide a variable for that base directory in the future.

Reply:
Done in https://github.com/rust-lang/rfcs/pull/3371/commits/b81b3edea0353fcf63568eac84a5e95360154459

Reply:
I'm not seeing how that commit addresses this.

Reply:
Misunderstanding, better done in https://github.com/rust-lang/rfcs/commit/5dcfb2906a337605b6c25aa7a9d743462828b6a5